src/gui/editorwidgets/qgskeyvaluewidgetfactory.cpp

```diff
@@ -43,5 +43,41 @@ QgsEditorConfigWidget *QgsKeyValueWidgetFactory::configWidget( QgsVectorLayer *v
 unsigned int QgsKeyValueWidgetFactory::fieldScore( const QgsVectorLayer *vl, int fieldIdx ) const
 {
   const QgsField field = vl->fields().field( fieldIdx );
+  if ( field.type() == QMetaType::Type::QVariantMap && ( field.typeName().compare( QStringLiteral( "JSON" ), Qt::CaseSensitivity::CaseInsensitive ) == 0 || field.subType() == QMetaType::Type::QString ) )
+  {
+    // Fetch the first not-null value and check if it is really a map
+    QgsFeatureRequest req;
+    req.setFlags( Qgis::FeatureRequestFlag::NoGeometry );
+    req.setSubsetOfAttributes( { fieldIdx } );
+    req.setFilterExpression( QStringLiteral( R"("%1" IS NOT NULL)" ).arg( field.name() ) );
```

Comment:
proper identifier escaping should likely be done for robustness. Same in other files.

I'm also wondering about if doing such feature request could not be sometimes problematic, thinking in particularly to slow datasources like WFS / OGC API Features, where if you are unlucky, it could result in pure client-side evaluation, and ingesting the whole layer if the field appends to be empty.

Reply:
That's a good point. I'll check if there is a way to limit this logic to datasources that don't have this problem.

Reply:
@rouault I couldn't find a capability for that, I changed the logic to NOT do the filtering and just examine the first 20 features (stopping at the first NOT NULL one).